numpy/core/arrayprint.py

```diff
@@ -1242,20 +1245,33 @@ def array_repr(arr, max_line_width=None, precision=None, suppress_small=None):
 
     if skipdtype:
         return "%s(%s)" % (class_name, lst)
+
+    # determine typename
+    if issubclass(arr.dtype.type, flexible):
+        if arr.dtype.names:
+            typename = "%s" % str(arr.dtype)
+        else:
+            typename = "'%s'" % str(arr.dtype)
     else:
         typename = arr.dtype.name
-        # Quote typename in the output if it is "complex".
+        # quote typenames which can't be represented as python variable names
         if typename and not (typename[0].isalpha() and typename.isalnum()):
             typename = "'%s'" % typename
 
-        lf = ' '
+    prefix = "{}({},".format(class_name, lst)
+    suffix = "dtype={})".format(typename)
+
+    # compute whether we should put dtype on a new line: Do so if adding the
+    # dtype would extend the last line past max_line_width.
+    last_line_len = len(prefix) - prefix.rfind('\n') - 1
```

Comment:
Is this correct when `\n` is not in `prefix, and `rfind` returns `-1`?

Reply:
I considered that, and I think it's ok:
```python
[1]: f = lambda x: len(x) - x.rfind("\n") - 1
[2]: f("AAA")
3
[3]: f("AA\nA")
1
[4]: f("AA\n")
0
```

Reply:
Huh, you're right. Would be a little clearer as `len(prefix) - (prefix.rfind('\n') + 1)`, and could probably do with a comment that -1 is actually being handled correctly.